### DISCUSSION TITLE
MP3 palyer, selection track...

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#15683] [Q&A] [answered]
The DFPlayer documentation in section 3.4.1, sections 2 and 3, states how a song command is constructed.
but, 
In section 3.4.4, example 2, states, that the format is 'folder' and 'song', so to select folder 02, song 003, it would be (2 x 256) + 3. or ( (2-1)  x 256) + 3. 

This is not what I get when I test this with my test song file... something odd is going on that I don't understand... can anyone help??

If I send cmd:  MP3Track 12, I get, folder 05, song 12 
~~~
09:37:42.080 CMD: mp3track 12
09:37:42.087 ---> Selection: 12, 0, 12
09:37:42.089 ---> Selection: 000c, 00, 0c
09:37:43.101 MQT: RSF/TASMOTA/6C4EFC/RESULT = {"MP3Track":12}
~~~

If I send cmd:  MP3Track 248, I get, folder 04, song 13 
~~~
09:34:27.920 CMD: mp3track 258
09:34:27.928 ---> Selection: 258, 1, 2
09:34:27.931 ---> Selection: 0102, 01, 02
09:34:28.944 MQT: RSF/TASMOTA/6C4EFC/RESULT = {"MP3Track":258}
~~~

![image](https://user-images.githubusercontent.com/4571691/170096993-962fd953-a226-4b29-bd7b-f7a801b52fc7.png)

**Comment by sfromis** [comment, marked as answer]
No experience with that thing, but https://wiki.dfrobot.com/DFPlayer_Mini_SKU_DFR0299 states:

> **NOTE:** The order you copy the mp3 into micro SD card will affect the order mp3 played , which means play(1) function will play the first mp3 copied into micro SD card.

**Comment by trlafleur** [comment]
the documentation is at odds with itself....  testing with a stand-alone
application I found, to see how it works....
I prefer to use the 'folder' 'song' format but will need to learn some more
about the device....

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Comment by trlafleur** [comment]
I discovered that there are two commands to the MP3 player to select a track, 0x03 will play a track in the root directory of the SD card, but 0x0f will play a song base on folder/song format.

For testing, I change the one byte for the track command to 0x0f and re-formatted my SD card...
For now, I can now send a folder as MSB of track command, (256, 512 ect + song number) to player. MP3Track 516, to play folder 2 song 4.

I will look at doing a pull request adding a new command -->  MP3Track1 ff,ss

**Reply by pkkrusty** [reply]
Nice work, this will add some handy functionality.

**Comment by trlafleur** [comment]
I am working on using my Seeburg Wallbox controller to build a complete jukebox with Tasmota...

I have a driver that now decodes the wallbox information and a Berry script that gets the information to pass on to the Tasmota MP3 player...

Issue: The current MP3 driver is a one-way device, it will not tell me when it's finished with a selection.
1) There are two ways to get busy from the DF-Player, implementing the RX serial pin to the driver and asking it, or adding the BUSY line to the driver...  I think the later is best...
2) How do you get Berry this information, I do not want to send an MQTT message out every second or so from the driver, as that adds lots of unneeded traffic.  

Question:  Is there a way for Berry to ask a driver for its state??

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/Commands/#mp3-player does not seem to include such. What you may or not be able to do with RX pin would be experimental, but seems a plausible thing to play around with.... Number of people here with deep experience with the thing may be "really low"....

For a bit more "advanced" usage with full control, I'd be tempted to redo the logic in Berry, seems to be relatively simple serial.

**Comment by trlafleur** [comment]
thanks, I now do understate how it works, well enough to make changes....

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
Now, you are the master of that device :smiley:

**Comment by trlafleur** [comment]
how can I from Berry get a status from the driver without going to MQTT...
is it as easy as setting a global variable and testing it from Berry?? is
there a way to do this from Berry??

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
Not sure how good a solution this is, but the [Rotary driver](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support_rotary.ino#L263) does "publish" JSON responses looking like they are for MQTT, but not actually published to MQTT. Can be picked up using `tasmota.add_rule`.

**Comment by trlafleur** [comment]
thanks for your help... basic stuff is working...



#-
To load this file, compile Tasmota32 with the option as needed....
Then Load the new binary image in your ESP32 and re-boot it.
Open the web page for this device, select Console, then Manage File System
Rename this Berry file to "autoexec.be", then upload it to the ESP32 file
system.
Reboot Tasmota, this Berry file will run after re-booting.
-#
#- *************************************** -#
#-
CHANGE LOG:
DATE REV DESCRIPTION
----------- --- ----------------------
28-Jan-2022 1.0 TRL - 1st release
Notes: 1) Tested with 11.1.0.3(tasmota)
ToDo: 1)

***@***.***


This is a jukebox player that get tracks to play from the Seeburg 3WA
wallbox via driver (xsns_123_SB-3Wx)
It requires a change to the MP3 driver (xdrv_14_mp3) to sense an external
busy siginal from the MP3 player.
this is made avilable to Berry as a rule.
-#

import string

#- *************************************** -#
#- *************************************** -#
class SEEBURG_DRIVER : Driver
#build an global array-->list to store wallbox events
static buf = []
var BusyFlag # 0 = busy
var TimeCount
var RandomPlay

#- *************************************** -#
def init()
tasmota.add_rule ("Wallbox", /MyObj -> self.process_wallbox(MyObj) )
tasmota.add_rule ("MP3Player", /MyObj2 -> self.process_MP3(MyObj2) )

self.BusyFlag = 0
self.RandomPlay = 0
end


#- *************************************** -#
def process_MP3(MyObj2)
#print (" In Process MP3")
#print("Obj2: ", MyObj2)
if MyObj2 == nil print("Bad Obj2") return end
if !(MyObj2.contains('MP3Busy')) return end
self.BusyFlag = real(MyObj2['MP3Busy'] )
#print("BusyFlag: ",self.BusyFlag)
end

#- *************************************** -#
def process_wallbox(MyObj)
var MaxQueSize = 4

#print (" In Process Wallbox")
#print("Obj: ", MyObj)
if MyObj == nil print("Bad Obj") return end
if !(MyObj.contains('Number_Index')) return end
if !(MyObj.contains('Letter_Index')) return end
if !(MyObj.contains('Selection_Index')) return end
var num = real(MyObj['Number_Index'] )
var let = real(MyObj['Letter_Index'] )
var inx = real(MyObj['Selection_Index'] )

if (self.buf.size() >= MaxQueSize) return end # do not add selection if full
print("Number Index: ", num, " Letter Index: ", let, " Selection Index: ",
inx)

self.buf.push(inx) # add selection to list
#print(self.buf)
end


#- *************************************** -#
def every_second()

if (self.BusyFlag == 1)
if (self.buf.size() == 0) return end # nothing to do, selection is empty
var NextTrack = self.buf.pop(0) # get next track
print ("Next Track", NextTrack)
var MyCmd = string.format("MP3Track %u", int (NextTrack))
print ("MyCmd: ", MyCmd)
tasmota.cmd(MyCmd)
end
end
#- *************************************** -#
def web_sensor()
#tasmota.web_send_decimal(msg)
end

#- *************************************** -#
def json_append()
#tasmota.response_append(msg)
end

end # end of class SEEBURG_Driver : Driver


#- *************************************** -#
SEEBURG_Driver = SEEBURG_DRIVER()
tasmota.add_driver(SEEBURG_Driver)

#- ************ The Very End ************* -#








~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Comment by trlafleur** [comment]
I have it all working as a system, MP3 driver, Wallbox Decoder, Berry
Jukebox.

will do a pull request for MP3 changes soon...

>
All files are posted on my GitHub account, waiting on PCB from china to
finish the project...
https://github.com/trlafleur/Tasmota_Seeburg_3WA_3W1_Decoder

>
Thanks for your help.



>
